software/contrib/menu.md

```diff
@@ -0,0 +1,142 @@
+# menu
+
+author: Matthew Jaskula
+
+date: 2022/02/15
+
+labels: utility
+
+A Menu which allows the user to select and execute one of the scripts available to this EuroPi. The menu will includes
+the contents of the contrib directory as well as the calibrate script. In addition, handlers are added that allow the
+user to exit the running script by rebooting the board back to this menu. Both the selection of a menu item and the
+exiting of a running program is achieved with the a 'hold one button and tap the other' action. This will work with 
+either button first.
+
+In the menu: 
+
+* button 1: scroll up
+* button 2: scroll down
+* tap button 1 while holding down button 2: launch selected program
+* tap button 2 while holding down button 1: launch selected program
+
+In a program that was launched from the menu:
+
+* tap button 1 while holding down button 2: exit the running program, returning to the menu
+* tap button 2 while holding down button 1: exit the running program, returning to the menu
+
+## Setup
+
+In order to use the menu you must copy it to ``main.py`` like any other script. In addition you must have all of the
+scripts you'd like to use copied to your pico in the places where the menu expects them. An example of the expected
+directory structure is below.
+
+```
+/
+├── main.py
+├── lib
+│   ├── contrib
+│   │   ├── coin_toss.py
+│   │   ├── diagnostic.py
+│   │   ├── harmonic_lfos.py
+│   │   ├── hello_world.py
+│   │   ├── menu.py
+│   │   ├── polyrhythmic_sequencer.py
+│   │   ├── radio_scanner.py
+│   │   └── scope.py
+│   ├── calibrate.py
+│   ├── europi.py
+│   └── ssd1306.py
+```
+
+You can use either Thony or rshell to copy the files to the pico. If you'd like to use rshell, the [deploy_firmware
+script](scripts/deploy_firmware.rshell) may be useful. It can be called using ``rshell -f scripts/deploy_firmware.rshell``
+
+## Menu Inclusion
+
+In order to be included in the menu a program needs to meet a few additional requirements, described below. Programs are
+not required to participate in the menu, but if they do not, they should be added to the ``EXCLUDED_SCRIPTS`` list near 
```

Comment:
Perhaps another possible approach would be to provide a base class that defines expected methods and behavior, then you can pull the list of script classes that implement the base class for the list of menu scripts instead of requiring users to declare exclusion.

https://docs.python.org/3/library/stdtypes.html?highlight=__subclasses__#class.__subclasses__

https://docs.python.org/3/library/stdtypes.html?highlight=__subclasses__#definition.__qualname__

Here's a drastically oversimplified explanation of my idea:
```python

class MenuScript:
  def main(self):
    """helpful documentation goes here"""
    raise NotImplementedError

class HelloWorld(MenuScript):
  def main(self):
    # properly implemented here

## in the menu.py
import MenuScript

# List scripts using their qualified name
for cls in MenuScript.__subclasses__():
    print(cls.__qualname__)
    # if selected, run that class
    asyncio.create_task(self._run_script(cls))

```

Reply:
Yeah, I thought about that, as mentioned in the original issue, and I do think that it would have a lot of benefits. I went in this direction because it would have the lowest impact on the existing scripts, and more closely matches how we are creating scripts today.

I expect that the use of the exclusion list would be very limited. This would be something that we check for when accepting submitted scripts. Either the script follows the requirements in this doc or it is added to the exclude list.

I can work up a PR with the base class idea if you like, so that we can see what it could look like. My suspicion would be that it's a little more complicated for not much more benefit.

Reply:
Sounds good to me. My gut says that a base class was the way to go, but the impression that I got from the issue's discussion was that we wanted this to be as few changes as possible.

Anyway, I look forward to seeing your base class implementation.

Reply:
Base class added